src/Microsoft.SqlTools.ServiceLayer/LanguageExtensibility/ExternalLanguageOperations.cs

```diff
@@ -0,0 +1,359 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using Microsoft.SqlTools.ServiceLayer.LanguageExtensibility.Contracts;
+using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+
+namespace Microsoft.SqlTools.ServiceLayer.LanguageExtensibility
+{
+
+    public enum ModifyType
+    {
+        Create,
+        Alter
+    }
+
+    public enum ContentModifyType
+    {
+        Modify,
+        Add,
+        Remove
+    }
+
+    public class ExternalLanguageOperations
+    {
+        private static string StatusScript = $@"
+SELECT is_installed
+FROM sys.dm_db_external_language_stats s, sys.external_languages l
+WHERE s.external_language_id = l.external_language_id AND language = @{LanguageNameParamName}";
+
+        private const string GetAllScript = @"
+SELECT l.external_language_id, language, l.create_date, dp.name, content, file_name, platform_desc, parameters, environment_variables
+FROM  sys.external_languages l 
+JOIN sys.external_language_files lf on l.external_language_id = lf.external_language_id
+JOIN sys.database_principals dp on l.principal_id = dp.principal_id
+ORDER BY l.external_language_id, platform";
+
+        private static string GetLanguageScript = $@"
+SELECT l.external_language_id, language, l.create_date, dp.name, content, file_name, platform_desc, parameters, environment_variables
+FROM  sys.external_languages l 
+JOIN sys.external_language_files lf on l.external_language_id = lf.external_language_id
+JOIN sys.database_principals dp on l.principal_id = dp.principal_id
+WHERE language=@{LanguageNameParamName}
+ORDER BY platform";
+
+        public const string CreateScript = "CREATE EXTERNAL LANGUAGE";
+        public const string DropScript = "DROP EXTERNAL LANGUAGE";
+        public const string AlterScript = "ALTER EXTERNAL LANGUAGE";
+        public const string SetContentScript = "SET";
+        public const string AddContentScript = "ADD";
+        public const string RemoveContentScript = "REMOVE";
+        private const string ContentParamName = "CONTENT";
+        private const string FileNameParamName = "FILE_NAME";
+        private const string PlatformParamName = "PLATFORM";
+        private const string EnvVariablesParamName = "ENVIRONMENT_VARIABLES";
+        private const string ParametersParamName = "PARAMETERS";
+        private const string LanguageNameParamName = "LANGUAGE";
+
+        private string GetDropScript(string languageName)
+        {
+            return $@"{DropScript} {languageName}";
```

Comment:
Shouldn't these need to be escaped properly?

Reply:
this is how TSQL support it

---
In reply to: [374371643](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r374371643) [](ancestors = 374371643)

Reply:
I haven't verified the syntax - but isn't the language-name an object name and thus can be bracket quoted? If that's the case you should do the bracket quoting and then escape the ] in the name - otherwise we leave ourselves open for sql injection.

Also - do you know if there's any plans to add support for external languages in SMO? That seems like a much better way to handle all this instead of having to write the queries ourselves...

---
In reply to: [374936060](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r374936060) [](ancestors = 374936060,374371643)

Reply:
yes we're planning to add this to SMO but I  couldn't wait for that. will fix the bracket 

---
In reply to: [375375572](https://github.com/microsoft/sqltoolsservice/pull/918#discussion_r375375572) [](ancestors = 375375572,374936060,374371643)